src/mixer/samplerbank.cpp

```diff
@@ -195,33 +195,48 @@ bool SamplerBank::loadSamplerBankFromPath(const QString& samplerBankPath) {
         return false;
     }
 
-    QDomNode n = root.firstChild();
-
-    while (!n.isNull()) {
+    const auto samplerNodes = root.childNodes();
+    if (samplerNodes.isEmpty()) {
+        return true;
+    }
+    for (int i = 0; i < samplerNodes.size(); i++) {
+        QDomNode n = samplerNodes.at(i);
         QDomElement e = n.toElement();
+        if (e.isNull() || e.tagName() != "sampler") {
+            continue;
+        }
+
+        const QString group = e.attribute("group", "");
+        const QString location = e.attribute("location", "");
+        int samplerNum;
+
+        if (group.isEmpty() || m_pPlayerManager->isSamplerGroup(group, &samplerNum)) {
+            continue;
+        }
+
+        // During startup we only increase the sampler count if there is
+        // a track to be loaded. This avoids
+        // * creating more (unneedded) sampler players than engine, skin or
+        //   some controller mapping requested
+        // * an unnecessary large Load To > Sampler N submenu in the track menu
+        if (initializing && location.isEmpty()) {
+            continue;
+        }
+
+        // Later on, when the user loads a samplers file manually, we
+        // want to eject loaded tracks if the file's sample slot is empty.
+        // We also create new players even if they are not present in the
+        // GUI to not drop tracks loaded to invisible samplers when saving
+        // the samplers file.
+        if (m_pPlayerManager->numSamplers() < (unsigned)samplerNum) {
```

Comment:
This way around should be more safe since samplerNum might be read from a corrupt file. 
```suggestion
        if (static_cast<int>(m_pPlayerManager->numSamplers()) < samplerNum) {
```

Reply:
Here is by the way also suspicious casting: 
https://github.com/mixxxdj/mixxx/blob/a7af5725c72804622fc44b270b634ca5e5b10b7b/src/mixer/playermanager.cpp#L278
Maybe we should just use int everywhere?

Reply:
You mean instead of `unsigned int`?

Reply:
Yes.

Reply:
Okay, I'll take a look and open another PR.